### ISSUE TITLE
Question about computation of precision and recall

### ISSUE THREAD

**Issue by noemi-mauro** [NicolasHug/Surprise#328]
Hi Nicholas,
I was looking at the computation of precision and recall 

        # Precision@K: Proportion of recommended items that are relevant
        precisions[uid] = n_rel_and_rec_k / n_rec_k if n_rec_k != 0 else 1

        # Recall@K: Proportion of relevant items that are recommended
        recalls[uid] = n_rel_and_rec_k / n_rel if n_rel != 0 else 1

and I was wondering why we assign 1 if n_rec_k  is 0 which means that there are not good recommendations. Isn't it wrong? This will increase the final mean over all the users.

Thank you,
Noemi

**Comment by NicolasHug** [comment]
I guess it's a convention. We could also have set those to 0.

In theroy the precision isn't defined if n_rec_k is 0, and the the recall isn't defined if n_rel is 0, so they should just be ignore.

**Comment by ashtou** [comment]
I agree with @noemi-mauro. Setting them to 1 is very confusing, plus, no warning is provided, so one might think that the system is working just fine, while it's totally the opposite.

**Comment by NicolasHug** [comment]
I'm happy to consider a PR replacing these with NaNs and use `np.nanmean` instead

**Comment by ashtou** [comment]
> I'm happy to consider a PR replacing these with `NaN`s and use `np.nanmean` instead

Ok, a few points here:

- Do you have an implementation reference that supports returning NaNs in such cases?

- I guess the best solution is provided by [SciKit](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_fscore_support.html) where they return 0 or 1 based on the value of the zero_division parameter `zero_division: 'warn', 0 or 1, default='warn'`. In the default case (`warn`), it acts as 0, but warnings are also raised. 

- I also think the default should be 0, and I have a strong case for that: suppose that we are comparing different recommender algorithms. In a specific case, one of them doesn't recommend any items, while the others do. Now, setting the precision of that algorithm to 1 is definitely misleading. Setting it to `NaN` and ignoring that case in the average (`np.nanmean`) is also misleading, because the case where the algorithm underperformed will be ignored. returning 0 with a warning seems like the most logical way to me.

**Comment by NicolasHug** [comment]
> Do you have an implementation reference that supports returning NaNs in such cases?

No. But it seems natural to me to use nan when a value isn't defined.

> I guess the best solution is provided by SciKit where they return 0 or 1 based on the value of the zero_division parameter zero_division: 'warn', 0 or 1, default='warn'

I would have doe that if precision/recall were a builtin metric, but they're not. Please keep in mind that what we're talking about is only a snippet in the FAQ to serve as an *example*, not as a reference. People using it should first read it, and adjust it to their needs. I believe the code is sufficiently clear and well documented such that the barrier is as low as possible.

There's no right or wrong solution here. Recall and precision aren't defined in these edge cases. Using 0 makes sense. Using nans makes sense. I'll give you that out of these, using 1 probably makes the least sense, but it's still a valid alternative.

**Comment by NicolasHug** [comment]
I'd be fine with using 0 too, with a comment at the top

**Comment by NicolasHug** [comment]
Fixed in #349 , closing
